### PR TITLE
Add finite differences gradcheck

### PR DESCRIPTION
**Pull request by szagoruyko**
Factored out torch_autograd.py testing code to a separate `torch.autograd.gradcheck` module. @apaszke please take a look.

address https://github.com/pytorch/pytorch/issues/849

Example:

```python
from torch.autograd import Variable, gradcheck

inputs = Variable(torch.randn(4, 4), requires_grad=True)

gradcheck(lambda x: 2*x.diag(), (inputs,), eps=1e-3)
```

### CODE REVIEW

File: test/test_autograd.py [hunk 1]
```diff
@@ -6,9 +6,9 @@
 import unittest
 from copy import deepcopy
 from collections import OrderedDict
+from torch.autograd import gradcheck
 
-from common import make_jacobian, TestCase, iter_tensors, \
```
Comment (fmassa) [inline test/test_autograd.py]: Do we still need to keep `make_jacobian` and `get_numerical_jacobian` in `common.py`? They are used as well in `common_nn.py`, but can't we use instead the ones from `autograd` that you just added?

File: torch/autograd/gradcheck.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            continue
+
+        def fn(input):
+            tmp = _as_tuple(func(*input))
+            return tmp[i].data
+
+        numerical = get_numerical_jacobian(fn, inputs, inputs, eps)
+        analytical = get_analytical_jacobian(_as_tuple(inputs), o)
+
+        if max(a.add(-1, n).abs().max()
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: can you remove a newline there? we allow up to 120 chars in line

File: test/test_autograd.py (hunk 1 above)
Comment (apaszke) [inline test/test_autograd.py]: Yes, we should use the one from autograd.

File: torch/autograd/gradcheck.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+def gradcheck(func, inputs, eps=1e-3, atol=1e-5):
+    """Check gradients computed via small finite differences
+       against analytical gradients
+
+    Args:
+        func: Python function that takes Variable inputs and returns
+            a tuple of Variables
+        inputs: tuple of Variables
+        eps: perturbation for finite differences
+        atol: absolute tolerance, if max abs violated return False
```
Comment (apaszke) [inline torch/autograd/gradcheck.py]: Why not just `tolerance`?

Comment (szagoruyko) [inline torch/autograd/gradcheck.py]: made it work as [numpy.allclose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html). has now both absolute and relative tolerance.

File: test/test_autograd.py (hunk 1 above)
Comment (szagoruyko) [inline test/test_autograd.py]: right, updated, thanks @fmassa @apaszke

**Review by apaszke — APPROVED** [review]

**Comment by soumith** [comment]
thanks Sergey!

**Pull request merged**